### PR TITLE
Kinno 611 send kin service module integration

### PR DESCRIPTION
**Pull request by shayb**
#### Main purpose:
made service to send kin
#### Technical description:
(What was changed/added)
#### Dilemmas you faced with? 
need to improve cashing data
#### Tests:
(Main tests you added/changed)
#### Documentation:
(Source/readme.md)

### CODE REVIEW

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java [hunk 1]
```diff
@@ -97,6 +97,13 @@ void signTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount,
 	void sendTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount, @NonNull String orderID,
 		@NonNull String offerID);
 
+	/**
+	 *
+	 * @param transactionEnvelope
+	 * @return
+	 */
+	String extractTransactionId(String transactionEnvelope);
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java:105]: can you explain on the comments here how?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java [hunk 2]
```diff
@@ -137,6 +139,23 @@ public void setMigrationManager(@NonNull final MigrationManager migrationManager
...
+		try {
+			kin.base.Transaction transaction = kin.base.Transaction.fromEnvelopeXdr(transactionEnvelope);
+			byte[] hashBytes = transaction.hash();
+			StringBuilder sb = new StringBuilder(hashBytes.length * 2);
+			for (byte b : hashBytes)
+				sb.append(String.format("%02x", b));
+			return sb.toString();
+		}
+		catch (IOException e) {
+			e.printStackTrace();
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java]: remove `printStackTrace `

File: core/src/main/java/com/kin/ecosystem/core/network/model/OutgoingTransferRequest.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+package com.kin.ecosystem.core.network.model;
+
+import java.util.Objects;
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * OutgoingTransferRequest
+ */
+public class OutgoingTransferRequest {
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/network/model/OutgoingTransferRequest.java]: I don't like the name as it not represent a request, only the payload.

File: core/src/main/java/com/kin/ecosystem/core/data/order/OrderRepository.java [hunk 4]
```diff
@@ -162,6 +163,13 @@ public void onFailure(ApiException e) {
 		});
 	}
 
+	@Override
+	public OpenOrder createOutgoingTransferOrderSync(@NonNull final OutgoingTransferRequest request) throws ApiException {
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OrderRepository.java]: this shouldn't throw ApiException it should throw `KinEcosystemException`.
Handle with try and catch and throw the relevant exception through `ErrorUtils.fromApiException`

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.TimerTask;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import kin.sdk.migration.common.exception.OperationFailedException;
+
+public class OutgoingTransferCall extends Thread {
+
+    private static final int SSE_TIMEOUT = 30 * 1000; // 30 seconds
+    private BlockchainSource blockchainSource = BlockchainSourceImpl.getInstance();
+    private OutgoingTransferRequest request;
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: it's not a request its a payload, please rename.

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import kin.sdk.migration.common.exception.OperationFailedException;
+
+public class OutgoingTransferCall extends Thread {
+
+    private static final int SSE_TIMEOUT = 30 * 1000; // 30 seconds
+    private BlockchainSource blockchainSource = BlockchainSourceImpl.getInstance();
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: pass the `BlockchainSource` on the ctor so you can add tests easily

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        this.callback = callback;
+        this.transferInfo = transferInfo;
+        this.isTimeoutTaskCanceled = new AtomicBoolean(false);
+        this.sseTimeoutTimer = new Timer();
+    }
+
+    @Override
+    public void run() {
+        try {
+            OrderRepository orderRepository = OrderRepository.getInstance();
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: Also, pass the `OrderRepository ` on the ctor.

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    private String transferInfo;
+    private final AtomicBoolean isTimeoutTaskCanceled;
+    private final Timer sseTimeoutTimer;
+    private Observer<Payment> paymentObserver;
+
+
+    OutgoingTransferCall(@NonNull OutgoingTransferRequest request, String transferInfo, @NonNull OutgoingTransferCallback callback) {
+        this.request = request;
+        this.callback = callback;
+        this.transferInfo = transferInfo;
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: what `transferInfo` represent? Can it be as an object?

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import java.math.BigDecimal;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import kin.sdk.migration.common.exception.OperationFailedException;
+
+public class OutgoingTransferCall extends Thread {
+
+    private static final int SSE_TIMEOUT = 30 * 1000; // 30 seconds
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: why 30sec? who decided this?

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    private boolean isTransferPayment(Payment payment) {
+        return request.getMemo().equals(payment.getOrderID());
+    }
+
+
+    interface OutgoingTransferCallback {
+
+        void onOutgoingTransferSuccess(OutgoingTransferRequest request, OpenOrder order, String transactionId);
+
+        void onOutgoingTransferFailed(OutgoingTransferRequest request, String errorMessage);
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: why we are passing only error msg ?
Can you pass the whole exception (KinEcosystemException)?

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            final OpenOrder order = orderRepository.createOutgoingTransferOrderSync(request);
+            payForOrder(orderRepository, order);
+        } catch (ApiException e) {
+            callback.onOutgoingTransferFailed(request, e.getMessage());
+        } catch (Exception e) {
+            callback.onOutgoingTransferFailed(request, e.getMessage());
+        }
+    }
+
+    private void payForOrder(final OrderRepository orderRepository, final OpenOrder order) {
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java:58]: you can remove `OrderRepository` after you pass it on the ctor.

File: core/src/main/java/com/kin/ecosystem/core/network/model/OutgoingTransferRequest.java (hunk 3 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/network/model/OutgoingTransferRequest.java]: @OrenZak  what would you call it in your name spaces

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java (hunk 9 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: after 30 sec on the client side we show to the user - that the transfer is delayed check balance later - no need to listen to block chain for response we will notify the user any more

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    @Override
+    public void run() {
+        try {
+            OrderRepository orderRepository = OrderRepository.getInstance();
+            final OpenOrder order = orderRepository.createOutgoingTransferOrderSync(request);
+            payForOrder(orderRepository, order);
+        } catch (ApiException e) {
+            callback.onOutgoingTransferFailed(request, e.getMessage());
+        } catch (Exception e) {
+            callback.onOutgoingTransferFailed(request, e.getMessage());
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: can we cancel the sse here ?

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                    orderRepository.submitSpendOrder(order.getOfferId(), transaction, order.getId(), transferInfo,
+                            new KinCallback<Order>() {
+                                @Override
+                                public void onResponse(Order order) {
+                                    scheduleTimeoutTimer();
+                                }
+
+                                @Override
+                                public void onFailure(KinEcosystemException exception) {
+                                    callback.onOutgoingTransferFailed(request, exception.getMessage());
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: can we cancel the sse here ?

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+				@Override
+				public void onOutgoingTransferSuccess(OutgoingTransferRequest request, OpenOrder order,
+													  String transactionId) {
+					callback.onSuccess(new KinTransferComplete(request.getWalletAddress(), transactionId, request.getMemo()));
+				}
+
+				@Override
+				public void onOutgoingTransferFailed(OutgoingTransferRequest request, String errorMessage) {
+					callback.onError(new KinTransferException(request.getWalletAddress(), errorMessage));
+				}
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java:50]: Why do we need a new Exception class?

Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java:50]: because this is the excpetion in the module side 
its not related to the ecosystem exceptions

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+			}).start();
+	}
+
+	@Override
+	public BigDecimal getCurrentBalance() throws BalanceException {
+		//need to get the updated balance every call not the cached
+		try {
+			return BlockchainSourceImpl.getInstance().getBalanceSync().getAmount();
+		} catch (Exception e) {
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java]: this creates a request to horizon, we can get from cache instead.

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+			throw new BalanceException(
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java]: This is new exception? why do we need it ?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java (hunk 1 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java:105]: done

File: core/src/main/java/com/kin/ecosystem/core/network/model/IncomingTransferRequest.java [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+package com.kin.ecosystem.core.network.model;
+
+import java.util.Objects;
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * IncomingTransferRequest
+ */
+
+public class IncomingTransferRequest {
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/network/model/IncomingTransferRequest.java]: Also doesn't like the name as it's not the actual request.

File: core/src/test/java/com/kin/ecosystem/core/data/order/OrderRepositoryTest.java [hunk 18]
```diff
@@ -184,7 +184,6 @@ public void createOrder_Failed() {
 		verify(openOrderCallback, never()).onResponse(any(OpenOrder.class));
 	}
 
-
```
Comment (OrenZak) [inline core/src/test/java/com/kin/ecosystem/core/data/order/OrderRepositoryTest.java:187]: nice

File: sdk/build.gradle [hunk 19]
```diff
@@ -32,16 +32,14 @@ dependencies {
     implementation devDependencies.appCompat_v7
     implementation devDependencies.recyclerView_v7
     implementation devDependencies.constraintLayout
+    //TODO need to update last version
+    implementation 'com.github.kinecosystem.move-kin-android:appsTransfer:202f69f95e0447def46cb4fb4cdbab40a035d637'
```
Comment (OrenZak) [inline sdk/build.gradle]: when it will be with the latest version and not a commit hash ?

File: sdk/src/main/java/com/kin/ecosystem/transfer/view/AccountInfoActivity.java [hunk 20]
```diff
@@ -18,13 +21,19 @@
     @Override
     public String getData() throws AccountInfoException {
         try {
+            //TODO how to do init if user is loggedin while activity came from background
+            //how will it be in kik - is there caching of address
+            Kin.initialize(getApplicationContext(), null);
```
Comment (OrenZak) [inline sdk/src/main/java/com/kin/ecosystem/transfer/view/AccountInfoActivity.java]: Lets talk about it.

**Review by OrenZak — CHANGES_REQUESTED** [review]
Please see my comments, can you add unit tests ?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java [hunk 21]
```diff
@@ -97,6 +97,14 @@ void signTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount,
 	void sendTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount, @NonNull String orderID,
 		@NonNull String offerID);
 
+	/**
+	 * Parse the envelope in order to retrieve the transaction hash
+	 * usage of  kin base low level
```
Comment (shayb) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java]: reprase

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java (hunk 9 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: consult with Hadar and Orad to change the timeout to 15 sec

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+    private final Timer sseTimeoutTimer;
+    private Observer<Payment> paymentObserver;
+    private OrderRepository orderRepository;
+    private BlockchainSource blockchainSource;
+
+
+    OutgoingTransferCall(@NonNull BlockchainSource blockchainSource, @NonNull OrderRepository orderRepository, @NonNull OutgoingTransfer payload, String transferInfo, @NonNull OutgoingTransferCallback callback) {
+        this.payload = payload;
+        this.callback = callback;
+        this.transferInfo = transferInfo;
```
Comment (shayb) [inline core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: change name to transferTitle

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java (hunk 16 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferService.java]: changed to KinEcosystemException

File: core/src/main/java/com/kin/ecosystem/core/network/model/IncomingTransferRequest.java (hunk 17 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/network/model/IncomingTransferRequest.java]: changed

File: sdk/build.gradle (hunk 19 above)
Comment (shayb) [reply sdk/build.gradle]: the coming days

File: core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java (hunk 9 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/order/OutgoingTransferCall.java]: change to 15 sec

File: core/src/main/java/com/kin/ecosystem/core/network/model/OutgoingTransferRequest.java (hunk 3 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/network/model/OutgoingTransferRequest.java]: `OutgoingTransferInfo`, `OutgoingTransferData`, `OutgoingTransfer` ?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java [hunk 23]
```diff
@@ -137,6 +139,21 @@ public void setMigrationManager(@NonNull final MigrationManager migrationManager
...
 	}
 
+	@Override
+	public String extractTransactionId(String transactionEnvelope) {
+		try {
+			kin.base.Transaction transaction = kin.base.Transaction.fromEnvelopeXdr(transactionEnvelope);
+			byte[] hashBytes = transaction.hash();
+			StringBuilder sb = new StringBuilder(hashBytes.length * 2);
+			for (byte b : hashBytes)
+				sb.append(String.format("%02x", b));
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java:149]: what is `%02x` ?

File: sdk/src/main/java/com/kin/ecosystem/Kin.java [hunk 24]
```diff
@@ -266,6 +266,31 @@ public void onFailure(KinEcosystemException exception) {
...
+		final BlockchainSource.Local local = BlockchainSourceLocal.getInstance(context);
+		final BlockchainSource.Remote remote = BlockchainSourceRemote
+				.getInstance(getInstance(context).executorsUtil, eventLogger);
+		MigrationManager migrationManager = new MigrationManager(context, appId, migrationNetworkInfo,
+				new KinBlockchainVersionProvider(local, remote),
+				new MigrationEventsListener(EventLoggerImpl.getInstance()), KIN_ECOSYSTEM_STORE_PREFIX_KEY);
+		migrationManager.enableLogs(Logger.isEnabled());
+		return migrationManager;
+	}
+
```
Comment (OrenZak) [inline sdk/src/main/java/com/kin/ecosystem/Kin.java:298]: why is that changed?

File: sdk/src/main/java/com/kin/ecosystem/transfer/view/AccountInfoWrapper.java [hunk 25]
```diff
@@ -0,0 +1,5 @@
+package com.kin.ecosystem.transfer.view;
+
+public class AccountInfoWrapper extends KinEcosystemWrapper {
```
Comment (OrenZak) [inline sdk/src/main/java/com/kin/ecosystem/transfer/view/AccountInfoWrapper.java]: do we need it ?

File: sdk/src/main/java/com/kin/ecosystem/transfer/view/KinEcosystemWrapper.java [hunk 26]
```diff
@@ -0,0 +1,15 @@
+package com.kin.ecosystem.transfer.view;
+
+import com.kin.ecosystem.base.KinEcosystemBaseActivity;
+
+public class KinEcosystemWrapper extends KinEcosystemBaseActivity {
```
Comment (OrenZak) [inline sdk/src/main/java/com/kin/ecosystem/transfer/view/KinEcosystemWrapper.java]: do we need it ?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java [hunk 27]
```diff
@@ -97,6 +97,14 @@ void signTransaction(@NonNull String publicAddress, @NonNull BigDecimal amount,
...
+	 * @param transactionEnvelope
```
Comment (OrenZak) [inline core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSource.java:102]: we didn't change that?

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java (hunk 23 above)
Comment (shayb) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java:149]: yossi code (stellar code)

File: sdk/src/main/java/com/kin/ecosystem/Kin.java (hunk 24 above)
Comment (shayb) [reply sdk/src/main/java/com/kin/ecosystem/Kin.java:298]: createMigrationManager is used in my class needed to access it

File: sdk/src/main/java/com/kin/ecosystem/transfer/view/AccountInfoWrapper.java (hunk 25 above)
Comment (shayb) [reply sdk/src/main/java/com/kin/ecosystem/transfer/view/AccountInfoWrapper.java]: no i will remove

File: sdk/src/main/java/com/kin/ecosystem/transfer/view/KinEcosystemWrapper.java (hunk 26 above)
Comment (shayb) [reply sdk/src/main/java/com/kin/ecosystem/transfer/view/KinEcosystemWrapper.java]: removed

File: core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java (hunk 23 above)
Comment (OrenZak) [reply core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java:149]: do we know what it does?
Can you ask him and add here a comment for later use?

File: sdk/src/main/java/com/kin/ecosystem/Kin.java (hunk 24 above)
Comment (OrenZak) [reply sdk/src/main/java/com/kin/ecosystem/Kin.java:298]: it can't be public this is the interface for the DS.
need to think of another solution.

Comment (shayb) [reply sdk/src/main/java/com/kin/ecosystem/Kin.java:298]: @OrenZak can i put this code as  private function in my class

**Review by OrenZak — APPROVED** [review]
💯
